elastic/beats
# Add k8s manifest leveraging leaderelection

## What does this PR do?
This PR proposes new Kubernetes manifests for shake of https://github.com/elastic/beats/issues/19731, leveraging unique Autodiscover provider implemented at https://github.com/elastic/beats/pull/20281.

With these manifests, only Metricbeat's Deamonset will be able to monitor whole k8s cluster since one Deamonset Pod each time will hold the leadership being responsible to coordinate metricsets that collect cluster wide metrics.

We will might need a meta issue to keep track of deprecating Deployment manifests (if needed). 

- [x] Add an extra section within https://www.elastic.co/guide/en/beats/metricbeat/7.x/running-on-kubernetes.html

## Why is it important?
To get rid of the requirement to maintain/handle two different deployment strategies.

## Checklist


## How to test this PR locally
Test the manifests similarly to the testing steps of https://github.com/elastic/beats/pull/20281#issue-457865292.

1. Prepare a multinode cluster (ie on GKE)
2. Edit `metricbeat-leaderelection-kubernetes.yml` properly to set the proper image (ie `docker.elastic.co/beats/metricbeat:7.10.0-SNAPSHOT`) and the proper ES output (ie on Elastic Cloud)
3. Deploy the `metricbeat-leaderelection-kubernetes.yml` manifest and make sure that all the desired metricsets are shipping events and that k8s related Dashboards are populated with data correctly.
4. `kubectl delete` the leader pod and make sure that the leadership is either transfered to another Pod or it is gained again by the new replacement Pod.
5. Add static or hints-based autodiscovery configs/providers and make sure that work all together.
Example:
```
metricbeat.autodiscover:
      providers:
        # To enable hints based autodiscover uncomment this:
        #- type: kubernetes
        #  node: ${NODE_NAME}
        #  hints.enabled: true
        - type: kubernetes
          node: ${NODE_NAME}
          templates:
            - condition:
                contains:
                  kubernetes.pod.name: "nats"
              config:
                - module: nats
                  hosts: "${data.host}:${data.port}"
        - type: kubernetes
          scope: cluster
          node: ${NODE_NAME}
          unique: true
          identifier: gke-lease
          templates:
            - config:
                - module: kubernetes
                  hosts: ["kube-state-metrics:8080"]
                  period: 10s
                  add_metadata: true
                  metricsets:
                    - state_node
                    - state_deployment
                    - state_replicaset
```

## Related issues
- Closes https://github.com/elastic/beats/issues/19731
